### PR TITLE
Provision internal URLs, CAs and proxy settings into dashboard

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
This PR about provisioning internal URLs, CAs and proxy settings into dashboard.
It's for https://github.com/eclipse-che/che-dashboard/pull/348

Note that dashboard will always use public Keycloak URL, since it needs user's info endpoint, and it gets it from Che Server.
Potentially it could evaluate it for itself, but I don't think it worthes the afford since:
- it's used only on Kubernetes and we expect it be dropped once we have native auth for Kubernetes;

### What issues does this PR fix or reference?
This is prereq for Dashboard part for https://github.com/eclipse/che/issues/20367

### How to test this PR?
Http proxy and custom certs are not easy to test, so I just copied Theia solution on dashboard side and assume it works.
1. For Che Self Signed certs verification you need to test Che on minikube:
```sh
cat EOF << > /tmp/che-cluster-patch.yaml
spec:
  k8s:
    singleHostExposureType: gateway
  server:
    dashboardImage: 'sleshchenko/che-dashboard:che-ca'
    dashboardImagePullPolicy: Always
    customCheProperties:
      CHE_INFRA_KUBERNETES_ENABLE__UNSUPPORTED__K8S: "true"
EOF
chectl server:deploy --installer=operator --platform=minikube \
  --che-operator-cr-patch-yaml=che-cluster-patch.yaml \
  --che-operator-image=sleshchenko/che-operator:dashboard-ca \
  --workspace-engine=dev-workspace
```
2. Open Dashboard and make sure dashboard reports the right values in logs:
Screenshot_20210916_152206

3. Disable internal URLs
```
oc patch checluster/eclipse-che -n eclipse-che --type=merge \
    --patch "{\"spec\":{\"server\":{\"disableInternalClusterSVCNames\":true}}}"
```
4. After Che operator did its job check Che Operator and Dashboard collaborate together well, and now public Che URL is used:
Screenshot_20210916_152924

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [Custom resource definition file is up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#updating-custom-resource-definition-file)
- [ ] [OLM bundles are up to date](https://github.com/eclipse-che/che-operator/blob/main/README.md#update-olm-bundle)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: pkg/deploy/dashboard/deployment_dashboard.go [hunk 1]
```diff
@@ -22,7 +23,43 @@ import (
...
+const CHE_CUSTOM_CERTS_MOUNT_PATH = "/public-certs/custom"
+
 func (d *Dashboard) getDashboardDeploymentSpec() (*appsv1.Deployment, error) {
+	var volumes []corev1.Volume
+	var volumeMounts []corev1.VolumeMount
+	var envVars []corev1.EnvVar
+
+	volumes, volumeMounts = d.provisionCustomPublicCA(volumes, volumeMounts)
+
+	selfSignedCertUsed, err := deploy.IsSelfSignedCertificateUsed(d.deployContext)
```
Comment (tolusha) [inline pkg/deploy/dashboard/deployment_dashboard.go:36]: `deploy.IsSelfSignedCertificateUsed` is it pretty heavy check.
For your case just check existence `self-signed-certificated` secret

File: pkg/deploy/dashboard/deployment_dashboard.go [hunk 2]
```diff
@@ -22,7 +23,43 @@ import (
...
 
+const CHE_SELF_SIGNED_MOUNT_PATH = "/public-certs/che-self-signed"
+const CHE_CUSTOM_CERTS_MOUNT_PATH = "/public-certs/custom"
+
 func (d *Dashboard) getDashboardDeploymentSpec() (*appsv1.Deployment, error) {
+	var volumes []corev1.Volume
+	var volumeMounts []corev1.VolumeMount
+	var envVars []corev1.EnvVar
+
+	volumes, volumeMounts = d.provisionCustomPublicCA(volumes, volumeMounts)
```
Comment (tolusha) [inline pkg/deploy/dashboard/deployment_dashboard.go:34]: What if `deploy.CheAllCACertsConfigMapName` does not exist ?

Comment (sleshchenko) [reply pkg/deploy/dashboard/deployment_dashboard.go:34]: Is it really possible?

So far, on kubernetes/openshift I configu nothing custom and it's there, but empty:
```
> kc get configmap ca-certs-merged -o=yaml
apiVersion: v1
kind: ConfigMap
metadata:
  annotations:
    che.eclipse.org/included-configmaps: ""
  creationTimestamp: "2021-09-16T13:36:10Z"
  labels:
    app.kubernetes.io/component: che
    app.kubernetes.io/instance: che
    app.kubernetes.io/managed-by: che-operator
    app.kubernetes.io/name: che
    app.kubernetes.io/part-of: che.eclipse.org
  name: ca-certs-merged
  namespace: eclipse-che
  ownerReferences:
  - apiVersion: org.eclipse.che/v1
    blockOwnerDeletion: true
    controller: true
    kind: CheCluster
    name: eclipse-che
    uid: 771b1491-c400-49a4-8ee6-2665b4c6f6bd
  resourceVersion: "300040"
  uid: 47316f04-d919-4bf5-82ca-b24d898d1f3f
```

File: pkg/deploy/dashboard/deployment_dashboard.go (hunk 1 above)
Comment (sleshchenko) [reply pkg/deploy/dashboard/deployment_dashboard.go:36]: well, I would agree with you but I can't:
1. `IsSelfSignedCertificateUsed` first of all check is secret exists and if it does.
2. it's exactly the same logic that is used for Che Server, so it suggested to me that checking is SA secret exists is not enough.

But then I finally agree. Let me refactor it a bit and ask you take a look a new commit.

File: pkg/deploy/dashboard/deployment_dashboard.go (hunk 2 above)
Comment (sleshchenko) [reply pkg/deploy/dashboard/deployment_dashboard.go:34]: As far as I see keycloak, che server mounts it without any check if it exists.
Do you still suggest to make such check before mounting?

File: pkg/deploy/dashboard/deployment_dashboard.go (hunk 1 above)
Comment (sleshchenko) [reply pkg/deploy/dashboard/deployment_dashboard.go]: here we go https://github.com/eclipse-che/che-operator/pull/1073/commits/cfe5333badbd017b07fe9b64248d613abf6ce7ce

**Review by tolusha — APPROVED** [review]

**Pull request merged**